lib/location/method_gnss.c

```diff
@@ -110,39 +159,45 @@ void method_gnss_timeout_work_fn(struct k_work *item)
 	method_gnss_cancel();
 }
 
-int method_gnss_init(void)
-{
-	int err = nrf_modem_gnss_init();
-
-	if (err) {
-		LOG_ERR("Failed to initialize GNSS interface, error %d", err);
-		return -err;
-	}
-
-	err = nrf_modem_gnss_event_handler_set(method_gnss_event_handler);
-	if (err) {
-		LOG_ERR("Failed to set GNSS event handler, error %d", err);
-		return -err;
-	}
-
-	k_work_init(&gnss_fix_work, method_gnss_fix_work_fn);
-	k_work_init(&method_gnss_timeout_work, method_gnss_timeout_work_fn);
-
-	return 0;
-}
-
-int method_gnss_location_get(const struct loc_method_config *config)
+static void method_gnss_positioning_work_fn(struct k_work *work)
 {
 	int err = 0;
-
-	const struct loc_gnss_config *gnss_config = &config->gnss;
+	int tau;
+	int active_time;
+	enum lte_lc_system_mode mode;
+
+	struct k_work_args *work_data =
+		CONTAINER_OF(work, struct k_work_args, work_item);
+
+	const struct loc_gnss_config gnss_config = work_data->gnss_config;
+
+	lte_lc_system_mode_get(&mode, NULL);
+
+	/* Don't care about PSM if we are in GNSS only mode */
+	if(mode != LTE_LC_SYSTEM_MODE_GPS) {
+		int ret = lte_lc_psm_get(&tau, &active_time);
+		if (ret < 0) {
+			LOG_ERR("Cannot get PSM config: %d. Starting GNSS right away.", ret);
+		} else if ((tau >= 0) & (active_time >= 0)) {
```

Comment:
Is the condition ok? Seems to work as expected in Telia network (no PSM) and in DNA network (PSM after 60s)

Reply:
I think checking the active time should be enough to detect PSM, but this could be confirmed from Jani M.

The current implementation waits for the device to enter PSM if A-GPS or P-GPS is enabled. For GNSS to be able to work without interruptions, we should always wait that RRC connection has been released and LTE has entered PSM (if PSM is enabled) before starting GNSS, I think. Fetching of assistance data causes LTE to enter connected mode, but it's possible that LTE is in connected mode also when A-GPS or P-GPS is not used (depends mostly on the application design).

Reply:
As discussed with @tokangas, he will change the if to only observe active time.

Reply:
I've changed this so that only active time is now checked.